tests/utilities/test_apply_func.py

```diff
@@ -303,6 +303,60 @@ def fn(a, b):
     assert reduced is None
 
 
+def test_apply_to_collections_dataclass():
+    @dataclasses.dataclass
+    class Feature:
+        input_ids: torch.Tensor
+        segment_ids: np.ndarray
+
+        def __eq__(self, o: object) -> bool:
+            if not isinstance(o, Feature):
+                return NotImplemented
+            else:
+                return torch.equal(self.input_ids, o.input_ids) and np.equal(self.segment_ids, o.segment_ids).all()
+
+    @dataclasses.dataclass
+    class ModelExample:
+        example_ids: List[str]
+        feature: Feature
+        label: torch.Tensor
+        some_constant: int = dataclasses.field(init=False)
+
+        def __post_init__(self):
+            self.some_constant = 7
+
+        def __eq__(self, o: object) -> bool:
+            if not isinstance(o, ModelExample):
+                return NotImplemented
+            else:
+                return (
+                    self.example_ids == o.example_ids
+                    and self.feature == o.feature
+                    and torch.equal(self.label, o.label)
+                    and self.some_constant == o.some_constant
```

Comment:
these definitions are now duplicated. 
a small improvement could be to pull all definitions out of the `test_recursive_application_to_collection` test and provide them at module level to be used by all tests.
Do you think we should do this?

Reply:
of course, that would be better. I'll move them.